src/mlpack/core/cv/k_fold_cv_impl.hpp

```diff
@@ -0,0 +1,190 @@
+                              const CVBaseArgs&... args) :
+    Base(args...), k(k)
+{
+  if (k < 2)
+    throw std::invalid_argument("KFoldCV: k should not be less than 2");
+
+  Init(Base::ExtractDataArgs(args...));
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<typename... MLAlgorithmArgs>
+double KFoldCV<MLAlgorithm,
+               Metric,
+               MatType,
+               PredictionsType,
+               WeightsType>::Evaluate(const MLAlgorithmArgs&... args)
+{
+  return TrainAndEvaluate(args...);
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+MLAlgorithm& KFoldCV<MLAlgorithm,
+                     Metric,
+                     MatType,
+                     PredictionsType,
+                     WeightsType>::Model()
+{
+  if (modelPtr == nullptr)
+    throw std::logic_error(
+        "KFoldCV::Model(): attempted to access an uninitialized model");
+
+  return *modelPtr;
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<typename DataArgsTupleT,
+         typename>
+void KFoldCV<MLAlgorithm,
+             Metric,
+             MatType,
+             PredictionsType,
+             WeightsType>::Init(const DataArgsTupleT& dataArgsTuple)
+{
+  InitKFoldCVMat(std::get<0>(dataArgsTuple), xs);
+  InitKFoldCVMat(std::get<1>(dataArgsTuple), ys);
+
+  Base::AssertDataConsistency(xs, ys);
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<typename DataArgsTupleT,
+         typename,
+         typename>
+void KFoldCV<MLAlgorithm,
+             Metric,
+             MatType,
+             PredictionsType,
+             WeightsType>::Init(const DataArgsTupleT& dataArgsTuple)
+{
+  InitKFoldCVMat(std::get<0>(dataArgsTuple), xs);
+  InitKFoldCVMat(std::get<1>(dataArgsTuple), ys);
+  InitKFoldCVMat(std::get<2>(dataArgsTuple), weights);
+
+  Base::AssertDataConsistency(xs, ys, weights);
+}
+
+template<typename MLAlgorithm,
+         typename Metric,
+         typename MatType,
+         typename PredictionsType,
+         typename WeightsType>
+template<typename SourceType, typename DestinationType>
+void KFoldCV<MLAlgorithm,
+             Metric,
+             MatType,
+             PredictionsType,
+             WeightsType>::InitKFoldCVMat(const SourceType& source,
+                                          DestinationType& destination)
+{
+  const DestinationType& sourceAsDT = source;
+
+  binSize = sourceAsDT.n_cols / k;
+  trainingSubsetSize = binSize * (k - 1);
+
+  destination = arma::join_rows(sourceAsDT,
+      sourceAsDT.cols(0, trainingSubsetSize - binSize - 1));
```

Comment:
This approach ends up causing two copies of 'source': the first line allocates
'source.n_elem' elements and copies from source to sourceAsDT, but then
join_rows must allocate an entirely new region of memory and copy everything
from sourceAsDT.

So I think you could just do arma::join_rows(source, source.cols(...)), but I
have not tested that.  (This laptop is a little Chromebook that takes ~60-90
minutes to compile all of mlpack... so I have not even tried building this
branch... :))